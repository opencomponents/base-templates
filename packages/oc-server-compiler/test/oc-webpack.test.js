jasmine.DEFAULT_TIMEOUT_INTERVAL = 100000;

('use strict');

const path = require('path');
const api = require('../lib/oc-webpack/index.js');
const MemoryFS = require('memory-fs');

const {
  compiler: webpackCompiler,
  configurator: webpackConfigurator
} = require('../lib/oc-webpack');

test('module APIs', () => {
  expect(api).toMatchSnapshot();
});

test('webpack configurator with production=false', () => {
  const config = webpackConfigurator({
    stats: 'errors-only',
    dependencies: {},
    production: false,
    publishFileName: 'server.js',
    serverPath: '/path/to/server.js'
  });

  // clean paths
  const target = config.module.rules[0].use;
  target[0].loader = path.relative(__dirname, target[0].loader);
  target[0].options.presets[0][0] = path.relative(
    __dirname,
    target[0].options.presets[0][0]
  );
  target[0].options.plugins[0][0] = path.relative(
    __dirname,
    target[0].options.plugins[0][0]
  );
  delete config.logger;

  expect(config).toMatchSnapshot();
});

test('webpack configurator with production=true', () => {
  const config = webpackConfigurator({
    stats: 'errors-only',
    dependencies: {},
    production: true,
    publishFileName: 'server.js',
    serverPath: '/path/to/server.js'
  });

  // clean paths
  const target = config.module.rules[0].use;
  target[0].loader = path.relative(__dirname, target[0].loader);
  target[1].loader = path.relative(__dirname, target[1].loader);
  target[1].options.presets[0][0] = path.relative(
    __dirname,
    target[1].options.presets[0][0]
  );
  target[1].options.plugins[0][0] = path.relative(
    __dirname,
    target[1].options.plugins[0][0]
  );
  delete config.logger;

  expect(config).toMatchSnapshot();
});

test('webpack compiler', done => {
  const serverPath = path.join(
    __dirname,
    '../../../mocks/jade-component',
    'server.js'
  );

  const dest = path.join(serverPath, '../build/server.js');
  const sourceMapDest = path.join(serverPath, '../build/server.js.map');
  const config = webpackConfigurator({
    stats: 'errors-only',
    dependencies: { lodash: '' },
    production: false,
    publishFileName: 'server.js',
    serverPath
  });

  webpackCompiler(config, (error, data) => {
    const fs = new MemoryFS(data);
    const serverContentBundled = fs.readFileSync(dest, 'UTF8');
    expect(serverContentBundled).toMatchSnapshot();

    const sourceMapContentBundled = fs.readFileSync(sourceMapDest, 'UTF8');
    const sourceMapJson = JSON.parse(sourceMapContentBundled);
    sourceMapJson.sources[1] = '/path/to/component/server.js';
    expect(sourceMapJson).toMatchSnapshot();
    done();
  });
});

test('webpack compiler verbose', done => {
  const serverPath = path.join(
    __dirname,
    '../../../mocks/jade-component',
    'server.js'
  );

  const dest = path.join(serverPath, '../build/server.js');
  const loggerMock = { log: jest.fn() };
  const config = webpackConfigurator({
    logger: loggerMock,
    stats: 'verbose',
    dependencies: { lodash: '' },
    production: true,
    publishFileName: 'server.js',
    serverPath
  });

  webpackCompiler(config, (error, data) => {
    const fs = new MemoryFS(data);
    const serverContentBundled = fs.readFileSync(dest, 'UTF8');
    const consoleOutput = loggerMock.log.mock.calls[0][0];
    expect(serverContentBundled).toMatchSnapshot();
    expect(consoleOutput).toMatch(/Hash:(.*?)18176142b4438edd492c/);
    expect(consoleOutput).toMatch(/Entrypoint(.*?)main(.*?)=(.*?)server.js/);
    expect(consoleOutput).toMatch(/external \"lodash\"/);
    done();
  });
});

test('webpack compiler with fatal error', done => {
  const config = webpackConfigurator({
    stats: 'errors-only',
    dependencies: { lodash: '' },
    publishFileName: 'server.js',
    serverPath: path.join(
      __dirname,
      '../../../mocks/jade-component',
      'server.js'
    )
  });

  config.plugins.push(function() {
    this.plugin('run', (compiler, cb) =>
      cb('This is a fatal compilation error')
    );
  });

  webpackCompiler(config, (error, data) => {
    expect(error).toMatchSnapshot();
    done();
  });
});

test('webpack compiler with soft error', done => {
  const config = webpackConfigurator({
    stats: 'errors-only',
    dependencies: { lodash: '' },
    publishFileName: 'server.js',
    serverPath: path.join(__dirname, 'some/not/valid/path', 'server.js')
  });

  webpackCompiler(config, (error, data) => {
    expect(error).toContain(`Entry module not found: Error: Can't resolve`);
    done();
  });
});

test('webpack compiler with warning', done => {
  const loggerMock = { log: jest.fn() };
  const config = webpackConfigurator({
    logger: loggerMock,
    stats: 'normal',
    dependencies: { lodash: '' },
    publishFileName: 'server.js',
    serverPath: path.join(
      __dirname,
      '../../../mocks/jade-component',
      'server.js'
    )
  });

  config.plugins.push(function() {
    this.plugin('run', (compiler, cb) => cb.call(compiler));
    this.plugin('done', stats => stats.compilation.warnings.push('A warning'));
  });

  webpackCompiler(config, (error, data) => {
    expect(loggerMock.log.mock.calls[0][0]).toContain('A warning');
    expect(error).toBe(null);
    done();
  });
});
