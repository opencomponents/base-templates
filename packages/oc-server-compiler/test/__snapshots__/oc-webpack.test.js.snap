// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module APIs 1`] = `
Object {
  "compiler": [Function],
  "configurator": [Function],
  "webpack": [Function],
}
`;

exports[`webpack compiler 1`] = `
"module.exports =
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = \\"\\";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = \\"./mocks/jade-component/server.js\\");
/******/ })
/************************************************************************/
/******/ ({

/***/ \\"./mocks/jade-component/server.js\\":
/*!****************************************!*\\\\
  !*** ./mocks/jade-component/server.js ***!
  \\\\****************************************/
/*! exports provided: data */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \\"data\\", function() { return data; });
const _ = __webpack_require__(/*! lodash */ \\"lodash\\");

const data = (context, callback) => {
  return callback(null, {});
};

/***/ }),

/***/ \\"lodash\\":
/*!*************************!*\\\\
  !*** external \\"lodash\\" ***!
  \\\\*************************/
/*! no static exports found */
/***/ (function(module, exports) {

module.exports = require(\\"lodash\\");

/***/ })

/******/ });
//# sourceMappingURL=server.js.map"
`;

exports[`webpack compiler 2`] = `
Object {
  "file": "server.js",
  "mappings": "dummy",
  "names": Array [
    "_",
    "require",
    "data",
    "context",
    "callback",
  ],
  "sourceRoot": "",
  "sources": Array [
    "webpack/bootstrap",
    "/path/to/component/server.js",
    "external \\"lodash\\"",
  ],
  "sourcesContent": Array [
    " 	// The module cache
 	var installedModules = {};

 	// The require function
 	function __webpack_require__(moduleId) {

 		// Check if module is in cache
 		if(installedModules[moduleId]) {
 			return installedModules[moduleId].exports;
 		}
 		// Create a new module (and put it into the cache)
 		var module = installedModules[moduleId] = {
 			i: moduleId,
 			l: false,
 			exports: {}
 		};

 		// Execute the module function
 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);

 		// Flag the module as loaded
 		module.l = true;

 		// Return the exports of the module
 		return module.exports;
 	}


 	// expose the modules object (__webpack_modules__)
 	__webpack_require__.m = modules;

 	// expose the module cache
 	__webpack_require__.c = installedModules;

 	// define getter function for harmony exports
 	__webpack_require__.d = function(exports, name, getter) {
 		if(!__webpack_require__.o(exports, name)) {
 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
 		}
 	};

 	// define __esModule on exports
 	__webpack_require__.r = function(exports) {
 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
 		}
 		Object.defineProperty(exports, '__esModule', { value: true });
 	};

 	// create a fake namespace object
 	// mode & 1: value is a module id, require it
 	// mode & 2: merge all properties of value into the ns
 	// mode & 4: return value when already ns object
 	// mode & 8|1: behave like require
 	__webpack_require__.t = function(value, mode) {
 		if(mode & 1) value = __webpack_require__(value);
 		if(mode & 8) return value;
 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
 		var ns = Object.create(null);
 		__webpack_require__.r(ns);
 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
 		return ns;
 	};

 	// getDefaultExport function for compatibility with non-harmony modules
 	__webpack_require__.n = function(module) {
 		var getter = module && module.__esModule ?
 			function getDefault() { return module['default']; } :
 			function getModuleExports() { return module; };
 		__webpack_require__.d(getter, 'a', getter);
 		return getter;
 	};

 	// Object.prototype.hasOwnProperty.call
 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };

 	// __webpack_public_path__
 	__webpack_require__.p = \\"\\";


 	// Load entry module and return exports
 	return __webpack_require__(__webpack_require__.s = \\"./mocks/jade-component/server.js\\");
",
    "const _ = require('lodash');

export const data = (context, callback) => {
  return callback(null, {});
};
",
    "module.exports = require(\\"lodash\\");",
  ],
  "version": 3,
}
`;

exports[`webpack compiler verbose 1`] = `"module.exports=function(a){function b(d){if(c[d])return c[d].exports;var e=c[d]={i:d,l:!1,exports:{}};return a[d].call(e.exports,e,e.exports,b),e.l=!0,e.exports}var c={};return b.m=a,b.c=c,b.d=function(a,c,d){b.o(a,c)||Object.defineProperty(a,c,{enumerable:!0,get:d})},b.r=function(a){'undefined'!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(a,Symbol.toStringTag,{value:'Module'}),Object.defineProperty(a,'__esModule',{value:!0})},b.t=function(a,c){if(1&c&&(a=b(a)),8&c)return a;if(4&c&&'object'==typeof a&&a&&a.__esModule)return a;var d=Object.create(null);if(b.r(d),Object.defineProperty(d,'default',{enumerable:!0,value:a}),2&c&&'string'!=typeof a)for(var e in a)b.d(d,e,function(b){return a[b]}.bind(null,e));return d},b.n=function(a){var c=a&&a.__esModule?function(){return a['default']}:function(){return a};return b.d(c,'a',c),c},b.o=function(a,b){return Object.prototype.hasOwnProperty.call(a,b)},b.p='',b(b.s=0)}([function(a,b,c){'use strict';c.r(b),c.d(b,'data',function(){return e});const d=c(1),e=(a,b)=>b(null,{})},function(a){a.exports=require('lodash')}]);"`;

exports[`webpack compiler with fatal error 1`] = `"This is a fatal compilation error"`;

exports[`webpack configurator with production=false 1`] = `
Object {
  "devtool": "#source-map",
  "entry": "/path/to/server.js",
  "externals": Array [
    [Function],
    /\\^\\[a-z@\\]\\[a-z\\\\-\\\\/0-9\\\\\\.\\]\\+\\$/i,
  ],
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "../../../node_modules/babel-loader/lib/index.js",
            "options": Object {
              "babelrc": false,
              "cacheDirectory": true,
              "plugins": Array [
                Array [
                  "../../../node_modules/babel-plugin-transform-object-rest-spread/lib/index.js",
                ],
              ],
              "presets": Array [
                Array [
                  "../../../node_modules/babel-preset-env/lib/index.js",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "node": 6,
                    },
                  },
                ],
              ],
              "retainLines": true,
              "sourceMaps": true,
              "sourceRoot": "/path/to",
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "minimize": false,
  },
  "output": Object {
    "devtoolFallbackModuleFilenameTemplate": "[absolute-resource-path]?[hash]",
    "devtoolModuleFilenameTemplate": "[absolute-resource-path]",
    "filename": "server.js",
    "libraryTarget": "commonjs2",
    "path": "/path/to/build",
  },
  "plugins": Array [
    DefinePlugin {
      "definitions": Object {
        "process.env.NODE_ENV": "\\"development\\"",
      },
    },
  ],
  "stats": "errors-only",
  "target": "node",
}
`;

exports[`webpack configurator with production=true 1`] = `
Object {
  "devtool": "",
  "entry": "/path/to/server.js",
  "externals": Array [
    [Function],
    /\\^\\[a-z@\\]\\[a-z\\\\-\\\\/0-9\\\\\\.\\]\\+\\$/i,
  ],
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "../../infinite-loop-loader/index.js",
          },
          Object {
            "loader": "../../../node_modules/babel-loader/lib/index.js",
            "options": Object {
              "babelrc": false,
              "cacheDirectory": true,
              "plugins": Array [
                Array [
                  "../../../node_modules/babel-plugin-transform-object-rest-spread/lib/index.js",
                ],
              ],
              "presets": Array [
                Array [
                  "../../../node_modules/babel-preset-env/lib/index.js",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "node": 6,
                    },
                  },
                ],
              ],
              "retainLines": true,
              "sourceMaps": false,
              "sourceRoot": "/path/to",
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "minimize": false,
  },
  "output": Object {
    "devtoolFallbackModuleFilenameTemplate": "[absolute-resource-path]?[hash]",
    "devtoolModuleFilenameTemplate": "[absolute-resource-path]",
    "filename": "server.js",
    "libraryTarget": "commonjs2",
    "path": "/path/to/build",
  },
  "plugins": Array [
    BabelMinifyPlugin {
      "options": Object {
        "babel": Object {
          "transform": [Function],
        },
        "comments": /\\^\\\\\\*\\*!\\|@preserve\\|@license\\|@cc_on/,
        "exclude": undefined,
        "include": undefined,
        "minifyOpts": Object {},
        "minifyPreset": [Function],
        "parserOpts": Object {},
        "sourceMap": undefined,
        "test": /\\\\\\.js\\(\\$\\|\\\\\\?\\)/i,
      },
    },
    DefinePlugin {
      "definitions": Object {
        "process.env.NODE_ENV": "\\"production\\"",
      },
    },
  ],
  "stats": "errors-only",
  "target": "node",
}
`;
